Fix the failure caused during restore when messages.log is deleted

anjumfatima90 opened the pull request:
fixes #18395

tjwatson commented on dev/com.ibm.ws.kernel.boot_test/test/com/ibm/ws/kernel/boot/internal/framework/BasicLogProvider.java:
```diff
@@ -26,4 +26,8 @@ public void configure(Map<String, String> config,
 
     @Override
     public void stop() {}
+    
```
Why is this override here?

anjumfatima90 replied:
This is because it implements LogProvider. I will make the update method default then i won't have to override it.

tjwatson commented on dev/com.ibm.ws.logging.core/src/com/ibm/wsspi/logprovider/LogProvider.java:
```diff
@@ -54,4 +54,12 @@
+     * value pairs. 
+     * @param config
+     */
+    void update(Map<String, Object> config);
```
This is related to my question above about why is this method overriden.  I believe this package to be SPI and that LogProvider may be implemented by other product extensions (not part of Liberty itself).  As such I think this will need to be a default method.

tjwatson commented on dev/com.ibm.ws.kernel.boot.nested/src/com/ibm/ws/kernel/launch/internal/FrameworkManager.java:
```diff
@@ -262,6 +267,49 @@ public Void run() throws Exception {
+
+                        @Override
+                        public void restore() {
+                            ClassLoader loader = config.getFrameworkClassloader();
```
Can you get the logProvider class from the logProvider instance that is there at checkpoint time?  I am pretty sure it has to be the exact same class name that you would get below from `config.getKernelResolver().getLogProvider()`

With that said, I am unsure why you need to create a new instance of the logProvider vs. constructing a new instance and calling update on it.  If you are creating a new instance why would it not get initialized the same way as it would the first time it is created before in checkpoint?

tjwatson commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderImpl.java:
```diff
@@ -34,6 +34,12 @@ public void configure(Map<String, String> config, File logLocation, TextFileOutp
 
+    @Override
+    public void update(Map<String, Object> config) {
+        TrConfigurator.update(config);
```
It seems all implementations of this do the same thing.  Can we not just call these two static methods ourselves?  Would that simplify things so we don't have to update the LogProvider interface?

tjwatson commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LoggingConstants.java:
```diff
@@ -65,4 +65,6 @@
         DEFAULT, IMMEDIATE
     };
+
+    String CHECKPOINT_ENABLED = "CHECKPOINT_ENABLED";
```
is this used?

tjwatson commented on dev/com.ibm.ws.logging.hpel/src/com/ibm/ws/logging/internal/hpel/HpelLogProviderImpl.java:
```diff
@@ -46,6 +46,12 @@ public void configure(Map<String, String> config, File logLocation, TextFileOutp
 
+    @Override
+    public void update(Map<String, Object> config) {
+        TrConfigurator.update(config);
```
I see no updates to TrConfigurator or FFDCConfigurator that would use the updated key.  Not sure how this flows down.

anjumfatima90 replied:
TrConfigurator.update(Map<String, Object>) has a line `config.update(newConfig);` which calls `LogProviderConfigImpl.update(Map<String, Object> -> LogProviderConfigImpl.doCommonInit(Map, boolean))` where I made changes.

anjumfatima90 committed:
Fix the failure caused during restore when messages.log is deleted

Add a checkpoint hook in FrameworkManager to stop the log provider during prepare and then update it during restore

tjwatson commented on dev/com.ibm.ws.kernel.boot.nested/src/com/ibm/ws/kernel/launch/internal/FrameworkManager.java:
```diff
@@ -272,6 +278,31 @@ public Void run() throws Exception {
+                    return checkpointHook;
+                }
+
+            }, null);
```
Is a service.ranking property needed here to ensure the hooks prepare is called last and the restore is called first?  I would think the service.ranking should be Integer min value so that it gets sorted properly when CheckpointImpl calls it.

tjwatson commented on dev/com.ibm.ws.kernel.boot.nested/src/com/ibm/ws/kernel/launch/internal/FrameworkManager.java:
```diff
@@ -272,6 +278,31 @@ public Void run() throws Exception {
+                        @Override
+                        public void restore() {
+                            config.put(BootstrapConstants.RESTORE_ENABLED, "true");
+                            Map<String, Object> configMap = Collections.<String, Object> unmodifiableMap(new ReadOnlyFrameworkProperties(config));
```
Why wrapper the ReadOnlyFrameworkProperties as an unmodifiableMap?  Seems redundant?

tjwatson replied:
I worry that this update with the original configMap may override settings that could happen dynamically from:

com.ibm.ws.logging.internal.osgi.LoggingConfigurationService.updated(Dictionary)

Is there any way to make it such that the TrConfigurator and FFDCConfigurator updates only pay attention to the RESTORE_ENABLED key when present and ignore all other keys?  This way we could have a singleton Map that only contains RESTORE_ENABLED key and all other settings that are currently set stay the same?

tjwatson commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -700,9 +715,20 @@ public void stop() {
 
```
On prepare we call stop().  This seems to restoreSystemStreams.  On restore I assume the system streams go back to the state they were?

anjumfatima90 commented:
#build

anjumfatima90 committed:
Set service ranking to min value and update TrConfiguartor with a singleton map on restore

anjumfatima90 committed:
Register logger handler and capture system streams on restore

anjumfatima90 commented:
#build

anjumfatima90 committed:
Fix method arguments

anjumfatima90 commented:
#build

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -196,6 +200,8 @@ public LogProviderConfigImpl(Map<String, String> config, File logLocation, TextF
 
         wlpUsrDir = config.get("wlp.user.dir");
+
+        checkpoint = config.get(LoggingConstants.CHECKPOINT_PROPERTY_NAME) != null ? true : false;
```
It would be better to use our existing method to get the Boolean value here, to be consistent:
`checkpoint = LoggingConfigUtils.getBooleanValue(config.get(LoggingConstants.CHECKPOINT_PROPERTY_NAME), checkpoint );`

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -126,6 +126,10 @@
     /** Allow JSON from applications write directly to System.out/System.err */
     protected volatile boolean appsWriteJson = false;
 
+    private final boolean checkpoint;
```
- Initialize this to equal false.
- Variables should be in camelCase to match other variables in the class.

tjwatson replied:
> * Initialize this to equal false.

It is a final field, it cannot be initialized and changed later

> * Variables should be in camelCase to match other variables in the class.

`checkpoint` is a single word, not "check point".

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/utils/FileLogHolder.java:
```diff
@@ -142,7 +142,7 @@ private static TraceComponent getTc() {
     public static FileLogHolder createFileLogHolder(TraceWriter oldLog, FileLogHeader logHeader,
```
Update copyright header to 2021.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/LogProviderConfigImpl.java:
```diff
@@ -126,6 +126,10 @@
     /** Allow JSON from applications write directly to System.out/System.err */
```
Update copyright header to 2021.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/Jsr47TraceService.java:
```diff
@@ -56,13 +56,17 @@ protected void unregisterLoggerHandlerSingleton() {}
     protected void initializeWriters(LogProviderConfigImpl config) {
         if (!WsLogManager.isConfiguredByLoggingProperties()) {
             // createFileLog may or may not return the original log holder..
```
Update copyright header to 2021.

pgunapal commented on dev/com.ibm.ws.logging/src/com/ibm/ws/logging/internal/impl/BaseTraceService.java:
```diff
@@ -315,11 +319,18 @@ protected void unregisterLoggerHandlerSingleton() {
         LogProviderConfigImpl trConfig = (LogProviderConfigImpl) config;
+        checkpoint = trConfig.isCheckpoint();
+        restore = trConfig.isRestore();
+        if (isRestore()) {
```
Isn't it better to use the restore value you got from line 324, instead of calling `isRestore()` directly?

anjumfatima90 committed:
Update copyright header year

anjumfatima90 commented:
#build

tjwatson commented on dev/com.ibm.ws.kernel.boot.nested/src/com/ibm/ws/kernel/launch/internal/FrameworkManager.java:
```diff
@@ -272,6 +277,30 @@ public Void run() throws Exception {
             // Set the framework variables only if everything succeeded.
             systemBundleCtx = fwk.getBundleContext();
+
+            if (LaunchArguments.isBetaEdition()) {
```
We don't need this beta block here.  CheckpointHookFactory services only get used when the beta checkpoint feature is enabled.

tjwatson reviewed (approved):
I'm going to merge this into criu to make progress in other areas.  @anjumfatima90 please open a separate PR to address the remaining two comments here.
